### DISCUSSION TITLE
Trying to extend the Order model

### DISCUSSION THREAD

**Discussion by ogalbnafets** [OFFLINE-GmbH/oc-mall-plugin#1103] [Q&A] [answered]
Hello,

I am trying to extend the order model so visitors can give additional information (order_comment). I already added a column through migration.

In the boot method I listened to order.beforeCreate
```
    Event::listen('mall.order.beforeCreate', function ($order) {
        \Log::info('Order comment:'. post('order_comment'));
        $order->order_comment = post('order_comment');
    });
```

I am using the Mall Starter Theme and added the field in the checkout component
```
<div class="mall-form-control">
    <label for="order_comment">Weitere Angaben:</label>
    <input class="form-control" id="order_comment" type="text"
           placeholder="Leer"
           name="order_comment">
    <div data-validate-for="order_comment"></div>
</div>
```

But the value is not being saved. Maybe post doesn't work? I am a newbie in extending plugins, so I would really appreciate help :)

Thanks!!

**Comment by tobias-kuendig** [comment]
Try attaching your event to the Order model's beforeSave event (each model in Laravel fires this):

https://docs.octobercms.com/3.x/extend/system/models.html#extending-models

```php
Order::extend(function($model) {
    $model->bindEvent('model.beforeSave', function() use ($model) {
        \Log::info('Order comment:'. post('order_comment'));
        // Maybe also check if this is the expected url using request()->url()
        if (post('order_comment')) {
            $order->order_comment = post('order_comment');
        }
    });
});
```

In this case, make sure the event listener only runs when you expect it (when the order is created in the frontend) and not everytime it is saved.

**Reply by ogalbnafets** [reply]
Hello, thank you. I forgot to say, that in the Event-Log the post doesn't show up. 
The checkout component is working with ajax I think - so maybe I have to add a new step-component with a seperate form would be my guess?
Thanks!

**Reply by ogalbnafets** [reply]
Dear Tobias, I am trying to implement this now for a few days, is there any possibility to get in touch with you? I would appreciate this, as I need this for a client (you can also bill me on this :))

Thanks in advance :)

**Reply by tobias-kuendig** [reply]
@ogalbnafets I'm sorry, but I'm not available for direct support. Did you use the code I have posted above? Is the `post` value empty?

If so, look at the request that is sent in your dev tools. If the comment field is not included there, this is the issue to fix. If the field is part of the `<form>` on the checkout page, it must be included in the request to the server.

**Reply by ogalbnafets** [reply, marked as answer]
Hello! Yes I used it, but the missing form was a hint! The "normal" checkout component has no form, so the field was also not being submitted! So instead I am using now the quick-checkout and now it's in the post!
The beforeSave Event doesn't fire off in my case, the test log also doesn't report. And in beforeCreate, the changed attribute apparently doesn't save. I looked in the model code, and in beforeCreates, the model is still the cart model. But after the copy into the order model, afterCreate fires off - so I used that and now it's working!! I hope this is okay?
The next problem was, that the extension of the form fields didn't work like in the documentation of october. I came up with the trick to override the controller template and copied the partials in my plugin and added the order_comment - works like a charm now :) I think it's not perfect in case Mall gets updated but I can compare with the commits and update my partials if wanted.
I will post my little code if another person is searching for the solution in future:

```
    public function boot()
    {    
        Order::extend(function($model) {
            $model->addFillable('order_comment');
        });
    

       \OFFLINE\Mall\Controllers\Orders::extend(function($controller) {
            list($author, $plugin) = explode('\\', strtolower(get_class()));
            $partials_path = sprintf('$/%s/%s/partials/orders', $author, $plugin);
            $controller->addViewPath($partials_path);
        });
    

        Event::listen('backend.list.extendColumns', function ($widget) {

            if (!$widget->model instanceof \OFFLINE\Mall\Models\Order) {
                return;
            }
            
            $widget->addColumns([
                'order_comment' => [
                    'label' => 'Kommentar'
                ],
            ]);
        
        });

               
       Event::listen('mall.order.afterCreate', function ($order, $cart) {
            $order->order_comment = post('order_comment');
        });            
    }
```
Here is the source of the trick with overriding the controller partials: [https://octobertricks.com/tricks/override-a-controllers-view-file-locally-in-your-plugin](url)

Thanks Tobias for your help!
Stefan

The field
